waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+  (form-encode clean-hash))
+
+(defn get-idp-redirect
+  "Return Ring response for HTTP 302 redirect."
+  [idp-url saml-request relay-state]
+  (redirect
+    (str idp-url
+         "?"
+         (let [saml-request (str->deflate->base64 saml-request)]
+           (uri-query-str
+             {:SAMLRequest saml-request :RelayState relay-state})))))
+
+(defn certificate-x509
+  "Takes in a raw X.509 certificate string, parses it, and creates a Java certificate."
+  [x509-string]
+  (let [certificate-factory (java.security.cert.CertificateFactory/getInstance "X.509")
+        certificate-input-stream (io/input-stream (.getBytes x509-string))]
+    (.generateCertificate certificate-factory certificate-input-stream)))
+
+(defn jcert->public-key
+  "Extracts a public key object from a java cert object."
+  [java-cert-obj]
+  (.getPublicKey java-cert-obj))
+
+(defn saml-assertion-signature-valid?
+  "Checks that the SAML assertion has a valid signature."
+  [assertion idp-cert]
+  (when-let [signature (.getSignature assertion)]
+    (let [idp-pubkey (-> idp-cert certificate-x509 jcert->public-key)
+          public-creds (doto (new org.opensaml.xml.security.x509.BasicX509Credential)
+                         (.setPublicKey idp-pubkey))
+          validator (new org.opensaml.xml.signature.SignatureValidator public-creds)]
+      (try
+        (.validate validator signature)
+        true
+        (catch ValidationException ex
+          (log/warn "Signature NOT valid" (.getMessage ex))
+          false)))))
+
+(defn new-doc-builder
+  []
+  (let [doc (DocumentBuilderFactory/newInstance)]
+    (.setNamespaceAware doc true)
+    (.setFeature doc "http://xml.org/sax/features/external-general-entities" false)
+    (.setFeature doc "http://xml.org/sax/features/external-parameter-entities" false)
+    (.setFeature doc "http://apache.org/xml/features/nonvalidating/load-external-dtd" false)
+    (.setAttribute doc "http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit" "2000")
+    (.setAttribute doc "http://www.oracle.com/xml/jaxp/properties/totalEntitySizeLimit" "100000")
+    (.setAttribute doc "http://www.oracle.com/xml/jaxp/properties/maxParameterEntitySizeLimit" "10000")
+    (.setAttribute doc "http://www.oracle.com/xml/jaxp/properties/maxElementDepth" "100")
+    (.setExpandEntityReferences doc false)
+    (.newDocumentBuilder doc)))
+
+(defn str->inputstream
+  "Unravels a string into an input stream so we can work with Java constructs."
+  [unravel]
+  (java.io.ByteArrayInputStream. (.getBytes unravel charset-format)))
+
+(defn str->xmldoc
+  [parsable-str]
+  (let [document (new-doc-builder)]
+    (.parse document (str->inputstream parsable-str))))
+
+(defn xml-string->saml-resp
+  "Parses a SAML response (XML string) from IdP and returns the corresponding (Open)SAML Response object"
+  [xml-string]
+  (let [xmldoc (.getDocumentElement (str->xmldoc xml-string))
+        unmarshallerFactory (org.opensaml.Configuration/getUnmarshallerFactory)
```

Comment:
Please use the simple class name in the body and move the fully qualified name to the import.

Reply:
done